### PR TITLE
ephemeris display improvements

### PR DESCRIPTION
**Pull request by tpolecat**
Some minor improvements in the display of ephemeris information in the target editor.

- shows ephemeris resolution in minutes
- shows time + position a bteginning of scheduling block (if any)
- shows current position, and updates in real time

image

There is still work to do but this is an improvement.

### CODE REVIEW

**Comment by tpolecat** [comment]
Hm where is Travis?

**Comment by tpolecat** [comment]
Ok close and then re-open fixed it. Confidence high!

File: bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details2/EphemerisEditor.scala [hunk 1]
```diff
@@ -1,47 +1,164 @@
...
 
 class EphemerisEditor extends TelescopePosEditor with ReentrancyHack {
 
-  val df = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss z") <| { df =>
-    df.setTimeZone(TimeZone.getTimeZone("UTC"))
-  }
+  private[this] var target: Option[NonSiderealTarget] = None
 
-  val start, end, size = new JLabel <| { l =>
+  // TODO: stop the timer when the window closes, or something?
```
Comment (cquiroz) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details2/EphemerisEditor.scala]: Could this stop the JVM from terminating? If the timer is still running

**Comment by cquiroz** [comment]
I wonder if this should be replicated on the PIT to some extent? Or even better, reused

File: bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details2/EphemerisEditor.scala (hunk 1 above)
Comment (tpolecat) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details2/EphemerisEditor.scala]: We call `System.exit()` so no, but I should at least create it as a daemon. I was trying to figure how to stop it when the panel is hidden but I was never able to see the relevant events so I gave up (`ComponentListener` never seemed to do anything). Any ideas?

**Comment by tpolecat** [comment]
Yeah, and I need to pull the custom ephemeris file thing from the PIT. We're very close to the models being identical so I think we should be able to make the editors shareable soon.

**Comment by tpolecat** [comment]
Here is a [movie](https://www.dropbox.com/s/vjzcp5mesr06u8k/moon.mov?dl=0). Kind of cool.

File: bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details2/EphemerisEditor.scala (hunk 1 above)
Comment (swalker2m) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details2/EphemerisEditor.scala]: Also, the `target` variable will be accessed from multiple threads.  Should it be an `AtomicReference`?

Comment (tpolecat) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details2/EphemerisEditor.scala]: What is the current thinking on that? Back in the olden days I always assumed reference updates were atomic. Did the 64-bit JVM change that?

Comment (swalker2m) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details2/EphemerisEditor.scala]: I think there shouldn't be an issue with a partial write like you would see with updating a Double or Long, but my understanding was that to guarantee that changes made by one thread are seen by another you have to mark the reference `@volatile` or else synchronize it.

**Comment by cquiroz** [comment]
Cool, Can you put the movie directly into GH or do you need to go the dropbox route?

**Comment by cquiroz** [comment]
Cool, Can you put the movie directly into GH or do you need to go the dropbox route?

**Comment by tpolecat** [comment]
No, it won't let you drop movies into comments.

**Comment by cquiroz** [comment]
Looks good :+1:

**Comment by arturog8m** [comment]
Now they will want TPE to show an animation with the trajectory of the object....

**Comment by andrewwstephens** [comment]
I'm not ruling out an animation :smile: but it seems like now we should be able to display the current and future positions without having to click a "Plot" button.

File: bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details2/EphemerisEditor.scala [hunk 2]
```diff
@@ -1,47 +1,164 @@
...
+  private val timer = new Timer(true)
+  timer.scheduleAtFixedRate(new TimerTask {
+    def run(): Unit =
+      target.flatMap(_.coords(System.currentTimeMillis)) match {
+        case Some(cs) => now.setText(cs.ra.toAngle.formatHMS + " " + cs.dec.formatDMS)
+        case None     => now.setText("--")
```
Comment (swalker2m) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details2/EphemerisEditor.scala]: We should put the `now.setText` in the event loop with `Swing.onEDT` too.

Comment (tpolecat) [inline bundle/jsky.app.ot/src/main/scala/jsky/app/ot/gemini/editor/targetComponent/details2/EphemerisEditor.scala]: Good catch. I really need to stop trying to write UI code.

**Comment by tpolecat** [comment]
Ok I made the `target` field volatile, moved the text field update on to the EDT, and turned down the refresh speed from 50ms to 1000ms … it looks cool when it goes that fast but there's no reason to burn CPU on it.

**Comment by tpolecat** [comment]
Also made some of the flatmappy stuff easier to follow.

**Comment by swalker2m** [comment]
Looks great. :+1:

**Pull request merged**